File: Java/54.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        LinkedList<Integer> ans = new LinkedList<Integer>();
        
        int m = matrix.length;
        
        if( m == 0)
            return (List)(new LinkedList<Integer>());
        
        int n = matrix[0].length;
        
        int x = 0,y = 0;
        
        int d = Math.min(m,n)/2;
        
        if( m == 1)
        {
            for(;y<n;y++)
                ans.add(matrix[x][y]);
            
            return (List)ans;
        }
        
        if(n == 1)
        {
            for(;x<m;x++)
                ans.add(matrix[x][y]);
            
            return (List)ans;
        }
        
        
        int count = m*n;
        
        for(int i = 0; i< d; i++)
        {
             x = i;
             y = i;
            
            for(;y<(n-1-i);y++)
             {ans.add(matrix[x][y]);count--;}
            
            for(;x<(m-1-i);x++)
                {ans.add(matrix[x][y]);count--;}
            
            for(;y>i;y--)
                {ans.add(matrix[x][y]);count--;}
            
            for(;x>i;x--)
                {ans.add(matrix[x][y]);count--;}
        }
            
            x++;
            y++;
            int i = x;
            
            if(count==1)
                {ans.add(matrix[x][y]);count--;}
            
            
            if(m<n && count>0)
            {
                for(;y<(n-i);y++)
                    ans.add(matrix[x][y]);
            }
            
            if(m>n && count>0)
            {
                for(;x<(m-i);x++)
                    ans.add(matrix[x][y]);
            }
        
        return (List)ans;
    }
}
__________________________________________________________________________________________________
sample 34412 kb submission
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        if(matrix == null || matrix.length == 0)
            return list;
        
        int m = matrix.length;
        int n = matrix[0].length;
        boolean used[][] = new boolean[m][n];
        
        int a = 0;
        while(a <= m - 1 - a){
            for (int j = a; j < n - a && a < m && !used[a][j]; j++){                
                list.add(matrix[a][j]);
                used[a][j]=true;
            }            
            for (int i = a + 1; i < m - a && i < m && n-1-a>=0 && !used[i][n-1-a]; i++) {
                list.add(matrix[i][n-1-a]);
                used[i][n-1-a] = true;
            }
            for (int j = n - 2 - a; j >= a && m-1-a>=0 && !used[m-1-a][j]; j--){ 
                list.add(matrix[m - 1 - a][j]);
                used[m-1-a][j] = true;
            }
            for (int i = m - 2 - a; i >= a + 1 && a < n && !used[i][a]; i--){
                list.add(matrix[i][a]);
                used[i][a] = true;
            }
            a++;
        }
                
        return list;
    }
}
__________________________________________________________________________________________________

File: Python3/54.py
__________________________________________________________________________________________________
sample 24 ms submission
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if matrix==[] or matrix[0]==[]:
            return []
        else:
            row = len(matrix)-1
            col = len(matrix[0])-1
            def nextInd(i,j,row,col):
                if i<=row/2 and i-j<=1 and i+j<col:
                    #go right
                    return i,j+1
                elif j>=col/2 and i+j>=col and i-j<row-col:
                    #go down
                    return i+1,j
                elif i>row/2 and i-j>=row-col and i+j>row:
                    #go left
                    return i,j-1
                elif j<col/2 and i+j<=row and i-j>1 :
                    #go up
                    return i-1,j
            opt=[]
            i = 0
            j = 0
            for k in range((row+1)*(col+1)):
                print(i,j)
                opt.append(matrix[i][j])
                if k < (row+1)*(col+1)-1:
                    i,j = nextInd(i,j,row,col)
            return opt
__________________________________________________________________________________________________
sample 12616 kb submission
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n = len(matrix)  
        if n == 0:
            return []
        m = len(matrix[0])
        if m == 0:
            return[]
        result = []
        i,j = 0,0
        start_i,start_j = 0,0
        end_i,end_j = n-1,m-1
        while True:
            #right
            if start_j <= end_j:
                result+=([matrix[i][x] for x in range(start_j,end_j+1)])
                start_i+=1
                j = end_j
            else:
                break
            #down
            if start_i <= end_i:
                result+=[matrix[x][j] for x in range(start_i,end_i+1)]
                end_j -= 1
                i = end_i
            else:
                break
            # left
            if start_j <= end_j:
                result += [matrix[i][x] for x in reversed(range(start_j, end_j + 1))]
                end_i -= 1
                j = start_j
            else:
                break
            # up
            if start_i <= end_i:
                result += [matrix[x][j] for x in reversed(range(start_i, end_i + 1))]
                start_j += 1
                i = start_i
            else:
                break
        return result
__________________________________________________________________________________________________
